### ISSUE TITLE
Keycloak fails to start due to infinispan state transfer exception

### ISSUE THREAD

**Issue by kopvortex** [keycloak/keycloak#21092]
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the latest release

### Area

infinispan

### Describe the bug

Running into following exception when starting new keycloak container on k8s. 
```
2023-06-19 21:42:16,446 TRACE [org.infinispan.factories.ComponentRegistry] (keycloak-cache-init) Invoking org.infinispan.counter.impl.CounterModuleLifecycle@7e76b944.cacheStopped()
2023-06-19 21:42:16,446 TRACE [org.infinispan.factories.ComponentRegistry] (keycloak-cache-init) Invoking org.infinispan.tasks.impl.LifecycleCallbacks@60d67fbd.cacheStopped()
2023-06-19 21:42:16,446 TRACE [org.infinispan.factories.ComponentRegistry] (keycloak-cache-init) Invoking org.infinispan.jboss.marshalling.JbossMarshallingModule@2a2df0de.cacheStopped()
2023-06-19 21:42:16,446 TRACE [org.infinispan.factories.ComponentRegistry] (keycloak-cache-init) Invoking org.infinispan.server.hotrod.LifecycleCallbacks@5fc79f76.cacheStopped()
2023-06-19 21:42:16,446 TRACE [org.infinispan.factories.ComponentRegistry] (keycloak-cache-init) Invoking org.infinispan.multimap.impl.MultimapModuleLifecycle@60ae908e.cacheStopped()
2023-06-19 21:42:16,446 TRACE [org.infinispan.factories.ComponentRegistry] (keycloak-cache-init) Invoking org.infinispan.persistence.remote.LifecycleCallbacks@adf8238.cacheStopped()
2023-06-19 21:42:16,446 TRACE [org.infinispan.factories.ComponentRegistry] (keycloak-cache-init) Invoking org.infinispan.server.core.LifecycleCallbacks@7c48f27b.cacheStopped()
2023-06-19 21:42:16,446 TRACE [org.infinispan.notifications.cachemanagerlistener.CacheManagerNotifierImpl] (keycloak-cache-init) Invoking listener: org.infinispan.remoting.inboundhandler.GlobalInboundInvocationHandler@3bbe1c15 passing event EventImpl{type=CACHE_STOPPED, newMembers=null, oldMembers=null, localAddress=null, viewId=0, subgroupsMerged=null, mergeVie
w=false}
2023-06-19 21:42:16,447 TRACE [org.infinispan.factories.impl.BasicComponentRegistryImpl] (keycloak-cache-init) Changed status of org.infinispan.globalstate.GlobalConfigurationManager to FAILED
2023-06-19 21:42:16,447 ERROR [org.infinispan.CONFIG] (keycloak-cache-init) ISPN000660: DefaultCacheManager start failed, stopping any running components: org.infinispan.commons.CacheException: java.lang.InterruptedException
        at org.infinispan.statetransfer.StateTransferManagerImpl.waitForInitialStateTransferToComplete(StateTransferManagerImpl.java:243)
        at org.infinispan.cache.impl.CacheImpl.start(CacheImpl.java:1013)
        at org.infinispan.cache.impl.AbstractDelegatingCache.start(AbstractDelegatingCache.java:504)
        at org.infinispan.manager.DefaultCacheManager.wireAndStartCache(DefaultCacheManager.java:723)
        at org.infinispan.manager.DefaultCacheManager.createCache(DefaultCacheManager.java:669)
        at org.infinispan.manager.DefaultCacheManager.internalGetCache(DefaultCacheManager.java:558)
        at org.infinispan.manager.DefaultCacheManager.getCache(DefaultCacheManager.java:521)
        at org.infinispan.security.actions.GetCacheAction.run(GetCacheAction.java:26)
        at org.infinispan.security.actions.GetCacheAction.run(GetCacheAction.java:14)
        at org.infinispan.security.Security.doPrivileged(Security.java:56)
        at org.infinispan.globalstate.impl.SecurityActions.doPrivileged(SecurityActions.java:30)
        at org.infinispan.globalstate.impl.SecurityActions.getCache(SecurityActions.java:39)
        at org.infinispan.globalstate.impl.GlobalConfigurationManagerImpl.start(GlobalConfigurationManagerImpl.java:111)
        at org.infinispan.globalstate.impl.CorePackageImpl$2.start(CorePackageImpl.java:60)
        at org.infinispan.globalstate.impl.CorePackageImpl$2.start(CorePackageImpl.java:48)
        at org.infinispan.factories.impl.BasicComponentRegistryImpl.invokeStart(BasicComponentRegistryImpl.java:617)
        at org.infinispan.factories.impl.BasicComponentRegistryImpl.doStartWrapper(BasicComponentRegistryImpl.java:608)
        at org.infinispan.factories.impl.BasicComponentRegistryImpl.startWrapper(BasicComponentRegistryImpl.java:577)
        at org.infinispan.factories.impl.BasicComponentRegistryImpl$ComponentWrapper.running(BasicComponentRegistryImpl.java:808)
        at org.infinispan.factories.AbstractComponentRegistry.internalStart(AbstractComponentRegistry.java:357)
        at org.infinispan.factories.AbstractComponentRegistry.start(AbstractComponentRegistry.java:250)
        at org.infinispan.manager.DefaultCacheManager.internalStart(DefaultCacheManager.java:775)
        at org.infinispan.manager.DefaultCacheManager.start(DefaultCacheManager.java:743)
        at org.infinispan.manager.DefaultCacheManager.<init>(DefaultCacheManager.java:407)
        at org.keycloak.quarkus.runtime.storage.legacy.infinispan.CacheManagerFactory.startCacheManager(CacheManagerFactory.java:96)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.InterruptedException
        at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:385)
        at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2022)
        at org.infinispan.statetransfer.StateTransferManagerImpl.waitForInitialStateTransferToComplete(StateTransferManagerImpl.java:236)
        ... 28 more
```
Subsequent startup failure. 
```
2023-06-19 21:42:16,491 DEBUG [org.infinispan.quarkus.hibernate.cache.QuarkusInfinispanRegionFactory] (main) Stop region factory
2023-06-19 21:42:16,492 DEBUG [org.infinispan.quarkus.hibernate.cache.QuarkusInfinispanRegionFactory] (main) Clear region references
2023-06-19 21:42:16,533 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Failed to start server in (production) mode
2023-06-19 21:42:16,533 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) Error details:: java.lang.RuntimeException: Failed to start caches
        at org.keycloak.quarkus.runtime.storage.legacy.infinispan.CacheManagerFactory.getOrCreate(CacheManagerFactory.java:61)
        at org.keycloak.quarkus.runtime.storage.legacy.infinispan.CacheManagerFactory_3e2e78b5a5eee8303325d41faca0a80d7da888f7_Synthetic_ClientProxy.getOrCreate(Unknown Source)
        at org.keycloak.quarkus.runtime.storage.legacy.infinispan.QuarkusCacheManagerProvider.getCacheManager(QuarkusCacheManagerProvider.java:32)
        at org.keycloak.connections.infinispan.DefaultInfinispanConnectionProviderFactory.lazyInit(DefaultInfinispanConnectionProviderFactory.java:143)
        at org.keycloak.connections.infinispan.DefaultInfinispanConnectionProviderFactory.create(DefaultInfinispanConnectionProviderFactory.java:83)
        at org.keycloak.connections.infinispan.DefaultInfinispanConnectionProviderFactory.create(DefaultInfinispanConnectionProviderFactory.java:67)
        at org.keycloak.services.DefaultKeycloakSession.getProvider(DefaultKeycloakSession.java:271)
        at org.keycloak.models.sessions.infinispan.InfinispanSingleUseObjectProviderFactory.getSingleUseObjectCache(InfinispanSingleUseObjectProviderFactory.java:53)
        at org.keycloak.models.sessions.infinispan.InfinispanSingleUseObjectProviderFactory.postInit(InfinispanSingleUseObjectProviderFactory.java:77)
        at org.keycloak.quarkus.runtime.integration.QuarkusKeycloakSessionFactory.init(QuarkusKeycloakSessionFactory.java:105)
        at org.keycloak.quarkus.runtime.integration.jaxrs.QuarkusKeycloakApplication.createSessionFactory(QuarkusKeycloakApplication.java:41)
        at org.keycloak.services.resources.KeycloakApplication.startup(KeycloakApplication.java:125)
        at org.keycloak.quarkus.runtime.integration.QuarkusLifecycleObserver.onStartupEvent(QuarkusLifecycleObserver.java:37)
        at org.keycloak.quarkus.runtime.integration.QuarkusLifecycleObserver_Observer_onStartupEvent_b0e82415b143738dc1f986a5fa4668e83d0a5dea.notify(Unknown Source)
        at io.quarkus.arc.impl.EventImpl$Notifier.notifyObservers(EventImpl.java:326)
        at io.quarkus.arc.impl.EventImpl$Notifier.notify(EventImpl.java:308)
        at io.quarkus.arc.impl.EventImpl.fire(EventImpl.java:76)
        at io.quarkus.arc.runtime.ArcRecorder.fireLifecycleEvent(ArcRecorder.java:131)
        at io.quarkus.arc.runtime.ArcRecorder.handleLifecycleEvents(ArcRecorder.java:100)
        at io.quarkus.deployment.steps.LifecycleEventsBuildStep$startupEvent1144526294.deploy_0(Unknown Source)
        at io.quarkus.deployment.steps.LifecycleEventsBuildStep$startupEvent1144526294.deploy(Unknown Source)
        at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
        at io.quarkus.runtime.Application.start(Application.java:101)
        at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:110)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:70)
        at org.keycloak.quarkus.runtime.KeycloakMain.start(KeycloakMain.java:98)
        at org.keycloak.quarkus.runtime.cli.command.AbstractStartCommand.run(AbstractStartCommand.java:37)
        at picocli.CommandLine.executeUserObject(CommandLine.java:1939)
        at picocli.CommandLine.access$1300(CommandLine.java:145)
        at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2358)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2352)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2314)
        at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)
        at picocli.CommandLine$RunLast.execute(CommandLine.java:2316)
        at picocli.CommandLine.execute(CommandLine.java:2078)
        at org.keycloak.quarkus.runtime.cli.Picocli.parseAndRun(Picocli.java:94)
        at org.keycloak.quarkus.runtime.KeycloakMain.main(KeycloakMain.java:88)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at io.quarkus.bootstrap.runner.QuarkusEntryPoint.doRun(QuarkusEntryPoint.java:61)
        at io.quarkus.bootstrap.runner.QuarkusEntryPoint.main(QuarkusEntryPoint.java:32)
Caused by: java.util.concurrent.TimeoutException
        at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:204)
        at org.keycloak.quarkus.runtime.storage.legacy.infinispan.CacheManagerFactory.getOrCreate(CacheManagerFactory.java:59)
        ... 42 more
```

Configuration:

We use the default cache-ispn.yaml and the configuration works fine on 20.0.3

### Version

21.1.1

### Expected behavior

Keycloak should start without any infinispan error. 

### Actual behavior

Keycloak container fails to start due to infinispan error. 

### How to Reproduce?

Able to reproduce on version 21.1.1 with following config

KC_CACHE_STACK - kubernetes
KC_HEALTH_ENABLED - true
KC_METRICS_ENABLED - true
KC_PROXY - reencrypt

**Comment by sschu** [comment]
Can you provide exact steps to reproduce?

**Comment by abstractj** [comment]
@kopvortex as mentioned by @sschu providing the steps to reproduce is essential for us to proceed. Otherwise, we may close this issue.

**Comment by kopvortex** [comment]
We have set up Keycloak on EKS (Elastic Kubernetes Service) with the integration of JGroup and Infinispan. To replicate the issue, I initiated a Keycloak deployment and then initiated another deployment before the first one was fully operational. Keycloak container in the second deployment fails with error. 

I noticed following in logs with jgroup trace logs enabled. 
```
2023-06-20 01:40:33,821 WARN  [org.jgroups.protocols.TCP] (TQ-Bundler-7,keycloak-7b987d9f68-rtbgz-2491) JGRP000032: keycloak-7b987d9f68-rtbgz-2491: no physical address for e0a9c418-fc6d-4635-89d3-4b74f3a1601f, dropping message                                      
java.lang.NullPointerException
        at org.jgroups.protocols.FD_SOCK2.getPhysicalAddresses(FD_SOCK2.java:445)
        at org.jgroups.protocols.FD_SOCK2.connectTo(FD_SOCK2.java:395)
        at org.jgroups.protocols.FD_SOCK2.connectToNextPingDest(FD_SOCK2.java:376)
        at org.jgroups.protocols.FD_SOCK2.handle(FD_SOCK2.java:347)
        at org.jgroups.protocols.FD_SOCK2.handle(FD_SOCK2.java:31)
        at org.jgroups.util.ProcessingQueue.process(ProcessingQueue.java:55)
        at org.jgroups.util.ProcessingQueue.add(ProcessingQueue.java:35)
        at org.jgroups.protocols.FD_SOCK2.handleView(FD_SOCK2.java:364)
        at org.jgroups.protocols.FD_SOCK2.down(FD_SOCK2.java:227)
        at org.jgroups.protocols.FailureDetection.down(FailureDetection.java:149)
        at org.jgroups.protocols.VERIFY_SUSPECT2.down(VERIFY_SUSPECT2.java:84)
        at org.jgroups.protocols.pbcast.NAKACK2.down(NAKACK2.java:619)
        at org.jgroups.protocols.UNICAST3.down(UNICAST3.java:611)
        at org.jgroups.protocols.pbcast.STABLE.down(STABLE.java:260)
        at org.jgroups.protocols.pbcast.GMS.installView(GMS.java:676)
        at org.jgroups.protocols.pbcast.ServerGmsImpl.handleViewChange(ServerGmsImpl.java:66)
        at org.jgroups.protocols.pbcast.ParticipantGmsImpl.handleViewChange(ParticipantGmsImpl.java:112)
        at org.jgroups.protocols.pbcast.GMS.handle(GMS.java:991)
        at org.jgroups.protocols.pbcast.GMS.up(GMS.java:855)
        at org.jgroups.protocols.pbcast.STABLE.up(STABLE.java:246)
        at org.jgroups.protocols.UNICAST3.up(UNICAST3.java:470)
        at org.jgroups.protocols.pbcast.NAKACK2.deliverBatch(NAKACK2.java:1010)
        at org.jgroups.protocols.pbcast.NAKACK2.removeAndDeliver(NAKACK2.java:943)
        at org.jgroups.protocols.pbcast.NAKACK2.handleMessageBatch(NAKACK2.java:915)
        at org.jgroups.protocols.pbcast.NAKACK2.up(NAKACK2.java:747)
        at org.jgroups.protocols.VERIFY_SUSPECT2.up(VERIFY_SUSPECT2.java:119)
        at org.jgroups.protocols.FailureDetection.up(FailureDetection.java:193)
        at org.jgroups.protocols.FD_SOCK2.up(FD_SOCK2.java:202)
        at org.jgroups.protocols.MERGE3.up(MERGE3.java:288)
        at org.jgroups.protocols.Discovery.up(Discovery.java:314)
        at org.jgroups.protocols.RED.up(RED.java:119)
        at org.jgroups.protocols.TP.passBatchUp(TP.java:1204)
        at org.jgroups.util.MaxOneThreadPerSender$BatchHandlerLoop.passBatchUp(MaxOneThreadPerSender.java:289)
        at org.jgroups.util.SubmitToThreadPool$BatchHandler.run(SubmitToThreadPool.java:150)
        at org.jgroups.util.MaxOneThreadPerSender$BatchHandlerLoop.run(MaxOneThreadPerSender.java:278)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)
```
Seems related to this https://issues.redhat.com/browse/JGRP-2707

**Comment by sschu** [comment]
Then the solution would be to wait until Keycloak picks up the Infinispan version where this is fixed. Also see https://github.com/keycloak/keycloak/issues/21119#issuecomment-1600419889

**Comment by abstractj** [comment]
Related to the fix provided by @pruivo here https://github.com/keycloak/keycloak/issues/21064.

**Comment by abstractj** [comment]
@stianst added 22 milestone here because this is important for us to include in the upcoming release. When you are back, please also review https://github.com/keycloak/keycloak/issues/21064. I believe that must to be included in 22.

**Comment by pruivo** [comment]
> @pruivo from my limited knowledge the changes proposed here #21064 can solve this issue. Is that correct?

It can workaround the "problem" (although I don't believe the NPE causes any issue and the cluster is able to recover).

The user can replace the protocol `FD_SOCK2` with `FD_SOCK` (or remove it). With the change in #21064 the following custom stack can be used:

```xml
<jgroups>
   <stack name="my-stack" extends="kubernetes">
      <FD_SOCK stack.combine="REPLACE" stack.position="FD_SOCK2"/>
   </stack>
</jgroups>
```
And set `KC_CACHE_STACK="my-stack"` (assuming it is running in OpenShift)

**Comment by ahus1** [comment]
I've been testing @pruivo's workaround together with the Keycloak Operator and found that it needed an additional configuration to make it work, as the DNS discovery hostname wasn't recognized as before. 

See https://github.com/keycloak/keycloak-benchmark/pull/440 for the PR in the Keycloak Benchmark project. 


Changes to the Infinispan cache configuration configuration file: 

```
    <!-- Workaround for https://github.com/keycloak/keycloak/issues/21092 -->
    <jgroups>
        <stack name="kubernetes-with-fdsock" extends="kubernetes">
            <!-- When using an embedded stack, replacement is done by Infinispan which requires environment variables to be
            prefixed with ".env", while JGroups falls back to environment variables without the prefix.
            The Keycloak Operator passes this information in an enviornment variable `jgroups.dns.query`.
            See https://github.com/keycloak/keycloak/issues/21830 for a discussion.
            -->
            <dns.DNS_PING dns_query="${env.jgroups.dns.query}" />
            <!-- Workaround for problems with FD_SOCK2, which might fail when the other nodes are not ready yet.
            See https://github.com/keycloak/keycloak/issues/21092 -->
            <FD_SOCK stack.combine="REPLACE" stack.position="FD_SOCK2"/>
        </stack>
    </jgroups>
```

And then adding the new cache name in the additional options in the Keycloak CR: 

```
  additionalOptions:
     # Workaround for https://github.com/keycloak/keycloak/issues/21092
    - name: 'cache-stack'
      value: kubernetes-with-fdsock
```

**Comment by ahus1** [comment]
One more question towards @kopvortex's comment above: 

> To replicate the issue, I initiated a Keycloak deployment and then initiated another deployment before the first one was fully operational. Keycloak container in the second deployment fails with error.

When deploying Keycloak with a StatefulSet, the first Pod should be ready before starting the second Pod. Among other reasons as outlined in https://github.com/keycloak/keycloak/discussions/11763#discussioncomment-2949975, this is one of the reasons why the Keycloak Operator is using a StatefulSet. 

Am I correct to assume you're using a Deployment for Keycloak in the scenario you described above?

**Comment by souravs17031999** [comment]
@ahus1, For me I am using StatefulSets and the scenario is like we have deployed Keycloak in k8s and two pods are running on v21, now we start to upgrade to v22, so first pod which comes up with v22 (keeping one of the older one at v21) starts to throw NPE exceptions and goes to crashLoop continuosly failing the deployment.

**Comment by sschu** [comment]
@souravs17031999 This is is not supported as the Infinispan versions in different Keycloak versions are not compatible. Furthermore, the new Keycloak version might contain database migrations. To do a Keycloak version upgrade, you have to scale to zero pods first and then update.

**Comment by souravs17031999** [comment]
Ok, thanks @sschu , makes sense.

**Comment by mabartos** [comment]
@ahus1 Thanks for improving and verifying the workaround.

To summarize what's the future of this issue:

We've been discussing it with the maintainers of JGroups and Infinispan, and we will be waiting for JGroups 5.2.17 and Infinispan 14.0.x releases. These releases will contain a fix for this issue. It might be included in the next planned micro-release 22.0.2 of Keycloak, or potentially in the next one.

**Comment by Jojoooo1** [comment]
@sschu would you mind pointing to any documentation that talk about the upgrade process and the need of scaling down to 0 (especially when using infinispan) ? I tried to find informations on this but was not able to find any. Thanks a lot for the help :)

**Comment by sschu** [comment]
This is not explicitely mentioned. You can infer this from the upgrading guide (https://www.keycloak.org/docs/latest/upgrading/index.html) because this talks about the traditional upgrade process of installed software which implicitly shuts down the software before upgrading it.

**Comment by ahus1** [comment]
@mabartos - I've started a draft PR targeting the latest 14.0.14-SNAPSHOT release, see #22386

**Comment by noiter** [comment]
I also have been following up a similar issue happened to me. One thing weird to me is that,
> no physical address for e0a9c418-fc6d-4635-89d3-4b74f3a1601f
The node name looks weird as other node names are like **keycloak-*** (keycloak-7b987d9f68-rtbgz-2491 in above example). Does anyone know why it?

**Comment by pruivo** [comment]
JGroups identifies nodes using UUID and it has a cache that maps UUID to logical names (`keycloak-*`) and physical addresses (IP and port). This cache is populated on demand and in your case, it seems the UUID is not present in the cache. The message should go away after a while after the message is retried and the cache is populated with that destination. 

If it does not go away, it may be some misconfiguration or some network issue.

**Comment by noiter** [comment]
Another issue I am facing when I am trying to override the kubernetes cache stack: two Keycloak instances started successfully, but same log `no members discovered after 2003 ms: creating cluster as coordinator` got printed for both instances. Looks like the distributed cache nodes failed to discover each other.

My custom cache stack config file is like,
```
<infinispan ...>
    <jgroups>
        <stack name="kubernetes-with-fdsock" extends="kubernetes">
            <dns.DNS_PING dns_query="${env.jgroups.dns.query}"/>
            <FD_SOCK stack.combine="REPLACE" stack.position="FD_SOCK2"/>
        </stack>
    </jgroups>

    <cache-container name="keycloak" statistics="true">
        <transport lock-timeout="60000" stack="kubernetes-with-fdsock"/>
        ... ...
    </cache-container>
</infinispan>
``` 
I have also tried
1. 
```
 <jgroups>
    <stack name="kubernetes-with-fdsock" extends="kubernetes">
        <FD_SOCK stack.combine="REPLACE" stack.position="FD_SOCK2"/>
    </stack>
</jgroups>
```
2. 
```
 <jgroups>
    <stack name="kubernetes-with-fdsock" extends="kubernetes">
        <dns.DNS_PING dns_query="${env.jgroups.dns.query}"/>
        <FD_SOCK stack.combine="REPLACE" stack.position="FD_SOCK2"/>
    </stack>
</jgroups>
```
3. 
```
 <jgroups>
    <stack name="kubernetes-with-fdsock" extends="kubernetes">
        <dns.DNS_PING dns_query="keycloak-headless"/>
        <FD_SOCK2 stack.combine="REMOVE"/>
    </stack>
</jgroups>
```
4. 
```
 <jgroups>
    <stack name="kubernetes-with-fdsock" extends="kubernetes">
        <dns.DNS_PING dns_query="keycloak-headless"/>
        <FD_SOCK stack.combine="REPLACE" stack.position="FD_SOCK2"/>
    </stack>
</jgroups>
```
My Keycloak config is like:
```
- name: KC_CACHE
  value: ispn
- name: KC_CACHE_CONFIG_FILE
  value: cache-ispn-fdsock.xml
- name: JAVA_OPTS_APPEND
  value: -Djgroups.dns.query=keycloak-headless ...
```

Can someone shed some lights on it? Thanks in advance.

**Comment by pruivo** [comment]
The first one should work but you need to configure the `cache-stack`. 
```
     # Workaround for https://github.com/keycloak/keycloak/issues/21092
    - name: 'cache-stack'
      value: kubernetes-with-fdsock
```
Example: https://github.com/keycloak/keycloak-benchmark/commit/880e4ea05a2262f0dc70861f6c608f17028325ad

**Comment by noiter** [comment]
@pruivo thanks for replying. I also tried it before, just tried it again, but unfortunately I got error below,
```
2023-08-21T12:26:06.527Z ERROR keycloak-cache-init org.infinispan.CONFIG : ISPN000660: DefaultCacheManager start failed, stopping any running components: org.infinispan.commons.CacheConfigurationException: ISPN000540: No such JGroups stack 'kubernetes-with-fdsock'
	at org.infinispan.configuration.global.JGroupsConfiguration.lambda$configurator$2(JGroupsConfiguration.java:69)
	at java.base/java.util.Optional.orElseThrow(Optional.java:403)
	at org.infinispan.configuration.global.JGroupsConfiguration.configurator(JGroupsConfiguration.java:69)
	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.buildChannel(JGroupsTransport.java:749)
	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.initChannel(JGroupsTransport.java:504)
	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.start(JGroupsTransport.java:485)
	at org.infinispan.remoting.transport.jgroups.CorePackageImpl$1.start(CorePackageImpl.java:42)
	at org.infinispan.remoting.transport.jgroups.CorePackageImpl$1.start(CorePackageImpl.java:27)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.invokeStart(BasicComponentRegistryImpl.java:616)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.doStartWrapper(BasicComponentRegistryImpl.java:607)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.startWrapper(BasicComponentRegistryImpl.java:576)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl$ComponentWrapper.running(BasicComponentRegistryImpl.java:807)
	at org.infinispan.metrics.impl.MetricsCollector.start(MetricsCollector.java:78)
	at org.infinispan.metrics.impl.CorePackageImpl$1.start(CorePackageImpl.java:41)
	at org.infinispan.metrics.impl.CorePackageImpl$1.start(CorePackageImpl.java:34)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.invokeStart(BasicComponentRegistryImpl.java:616)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.doStartWrapper(BasicComponentRegistryImpl.java:607)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.startWrapper(BasicComponentRegistryImpl.java:576)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl$ComponentWrapper.running(BasicComponentRegistryImpl.java:807)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.startDependencies(BasicComponentRegistryImpl.java:634)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.doStartWrapper(BasicComponentRegistryImpl.java:598)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.startWrapper(BasicComponentRegistryImpl.java:576)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl$ComponentWrapper.running(BasicComponentRegistryImpl.java:807)
	at org.infinispan.factories.AbstractComponentRegistry.internalStart(AbstractComponentRegistry.java:379)
	at org.infinispan.factories.AbstractComponentRegistry.start(AbstractComponentRegistry.java:252)
	at org.infinispan.manager.DefaultCacheManager.internalStart(DefaultCacheManager.java:778)
	at org.infinispan.manager.DefaultCacheManager.start(DefaultCacheManager.java:746)
	at org.infinispan.manager.DefaultCacheManager.<init>(DefaultCacheManager.java:410)
	at org.keycloak.quarkus.runtime.storage.legacy.infinispan.CacheManagerFactory.startCacheManager(CacheManagerFactory.java:96)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
: org.infinispan.commons.CacheConfigurationException: ISPN000540: No such JGroups stack 'kubernetes-with-fdsock'
	at org.infinispan.configuration.global.JGroupsConfiguration.lambda$configurator$2(JGroupsConfiguration.java:69)
	at java.base/java.util.Optional.orElseThrow(Optional.java:403)
	at org.infinispan.configuration.global.JGroupsConfiguration.configurator(JGroupsConfiguration.java:69)
	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.buildChannel(JGroupsTransport.java:749)
	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.initChannel(JGroupsTransport.java:504)
```

Any other suggestions?

**Comment by pruivo** [comment]
did you upload `cache-ispn-fdsock.xml` as a configmap?

**Comment by noiter** [comment]
I am using https://github.com/codecentric/helm-charts/tree/master/charts/keycloakx to do the Keycloak deployment. I don't think it is uploading `cache-ispn-fdsock.xml` as a configmap. I think it is simply specify the cache config file via KC_CACHE_CONFIG_FILE variable. 
I am also putting these KC_CACHE variables under **extraEnv** of the container keycloak of StatefulSet definition.

**Comment by pruivo** [comment]
I'm not familiar with that helm chart. 
Make sure `/opt/keycloak/conf/cache-ispn-fdsock.xml` file exists and it contains the stack with the name `kubernetes-with-fdsock`.

**Comment by noiter** [comment]
Yeah I also confirmed that but unfortunately it didn't work still. 

I've also tried locally with docker-compose, but as long as I specify the KC_CACHE_STACK with the custom stack name, it always throws error 
```
ISPN000540: No such JGroups stack 'tcp-k8s'
```
, but such error is never thrown when I leave KC_CACHE_STACK empty but follow the way described https://www.keycloak.org/server/caching#_custom_transport_stacks to configure. But problem with it that Infinispan nodes don't discovery each other. :(

**Comment by MrDWilson** [comment]
> @souravs17031999 This is is not supported as the Infinispan versions in different Keycloak versions are not compatible. Furthermore, the new Keycloak version might contain database migrations. To do a Keycloak version upgrade, you have to scale to zero pods first and then update.

I'm experiencing the same issue, and this sorted it for me. However, do we know if this is for major, minor or patch versions? We're going from 22 to 24 when noticing it, but would like to know for future if doing even a patch version would need a scale down first (as this would cause downtime).

**Comment by ahus1** [comment]
> do we know if this is for major, minor or patch versions? 

For now, we only support rolling upgrades when you stay on the exact same versions (including the patch level). You usually do this to change startup configurations or memory settings. 

We would eventually support rolling upgrades on patch releases. We're currently discussing this. Once we have the right tests in place, and are sure we can guarantee it, we'll add this to the release notes and also to the Keycloak upgrade guide - https://www.keycloak.org/docs/latest/upgrading/index.html#_upgrading

**Comment by viniciusd** [comment]
Sorry for bumping this thread, but, going over its entirety, I understand fixes and workarounds all targeted version 22 or 23 of Keycloak.

If one is upgrading from an older version and just facing it (20->21), what is the workaround in v21 before upgrading to v22?

__Edit__
If you are facing the same issue, as a workaround to get v21 up and running, we are going with `<state-transfer await-initial-transfer="false"/>` in the `distributed-cache` block in our `cache-ispn.xml`
